kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/PaginatedScanFilesIteratorImpl.java

```diff
@@ -15,52 +15,148 @@
    *     pagination.
    * @param paginationContext The pagination context that carries pagination-related information,
    *     such as the maximum number of files to return in a page.
    */
   public PaginatedScanFilesIteratorImpl(
-      CloseableIterator<FilteredColumnarBatch> filteredScanFilesIter,
+      CloseableIterator<FilteredColumnarBatch> baseFilteredScanFilesIter,
       PaginationContext paginationContext) {
-    this.filteredScanFilesIter = filteredScanFilesIter;
+    this.baseFilteredScanFilesIter = baseFilteredScanFilesIter;
     this.pageSize = paginationContext.getPageSize();
   }
 
   @Override
   public Row getCurrentPageToken() {
-    throw new UnsupportedOperationException("Not implemented");
+    // TODO: change values for data validation here
+    return new PageToken(
+            lastReadLogFilePath,
+            lastReturnedRowIndex,
+            Optional.empty() /* sidecar file index */,
+            Meta.KERNEL_VERSION,
+            "fake/table/path" /* table path */,
+            -1 /* table version */,
+            -1 /* predicate hash */,
+            -1 /* log segment hash */)
+        .toRow();
   }
 
   @Override
   public boolean hasNext() {
-    throw new UnsupportedOperationException("Not implemented");
+    if (closed) {
+      throw new IllegalStateException("Can't call `hasNext` on a closed iterator.");
+    }
+    if (!currentBatch.isPresent()) {
+      prepareNext();
+    }
+    return currentBatch.isPresent();
+  }
+
+  private void prepareNext() {
+    if (currentBatch.isPresent()) return;
+    if (numScanFilesReturned >= pageSize) return;
+    if (!baseFilteredScanFilesIter.hasNext()) return;
+
+    final FilteredColumnarBatch batch = baseFilteredScanFilesIter.next();
+    checkArgument(batch.getFilePath().isPresent(), "file path doesn't exist!");
+    checkArgument(
+        batch.getPreComputedNumSelectedRows().isPresent(),
+        "pre-computed number of selected rows doesn't exist!");
+    if (!batch.getFilePath().get().equals(lastReadLogFilePath)) {
+      lastReadLogFilePath = batch.getFilePath().get();
+      lastReturnedRowIndex = -1;
+      logger.info("filePath {}", lastReadLogFilePath);
+    }
+    final long numSelectedAddFilesInBatch = batch.getPreComputedNumSelectedRows().get();
+    final long numRowsInBatch = batch.getData().getSize();
+
+    currentBatch = Optional.of(batch);
+    numScanFilesReturned += numSelectedAddFilesInBatch;
+    lastReturnedRowIndex += numRowsInBatch;
+
+    logger.info("total numScanFilesReturned: {}", numScanFilesReturned);
+    logger.info("numSelectedAddFilesInBatch: {}", numSelectedAddFilesInBatch);
+    logger.info("numRowsInBatch: {}", numRowsInBatch);
   }
 
   @Override
   public FilteredColumnarBatch next() {
-    throw new UnsupportedOperationException("Not implemented");
+    if (closed) {
+      throw new IllegalStateException("Can't call `next` on a closed iterator.");
```

Comment:
https://github.com/delta-io/delta/blob/2095e31a7ba44600a5d3318d790274d36ff3380b/kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/Preconditions.java#L73 use this utils?